src/charm.py

```diff
@@ -5,80 +5,160 @@
+            return
+        event.add_status(ops.ActiveStatus())
+
+    def _on_gocert_notify(self, event: ops.PebbleCustomNoticeEvent):
+        gocert_csrs = self._certificate_requests_table
+        assert gocert_csrs  # TODO: error checking
+        databag_csrs = self.tls.get_requirer_csrs()
+        for csr in databag_csrs:
+            if csr.csr not in [row.get("CSR", "") for row in gocert_csrs]:
+                # TODO: error checking and signing up to gocert
+                requests.post(
+                    url=f"https://{self._application_bind_address}:{self.port}/api/v1/certificate_requests",
+                    data=csr.csr,
+                    headers={"Content-Type": "text/plain", "Authorization": "Bearer {token}"},
+                    verify=f"{CHARM_PATH}/{CONFIG_MOUNT}/0/ca.pem",
+                )
+
+        gocert_csrs = self._certificate_requests_table
+        assert gocert_csrs  # TODO: error checking
+        for row in gocert_csrs:
+            gocert_csr = row.get("CSR")
+            # gocert_cert = row.get("Certificate")
+            for databag_csr in databag_csrs:
+                if not databag_csr != gocert_csr:
+                    continue
+                # TODO: if certificate in gocert does not match our provider side, update
+
+    ## Configure Dependencies ##
+    def _configure_gocert_config_file(self):
+        """Push the config file."""
+        logger.info("[GoCert] Configuring the config file.")
         try:
             self.container.pull("/etc/config/config.yaml")
+            logger.info("[GoCert] Config file already created.")
         except ops.pebble.PathError:
-            config_file = open("src/configs/config.yaml").read()
-            cert_file = open("src/configs/cert.pem").read()
-            key_file = open("src/configs/key.pem").read()
+            config_file = open("src/config/config.yaml").read()
             self.container.make_dir(path="/etc/config", make_parents=True)
             self.container.push(path="/etc/config/config.yaml", source=config_file)
-            self.container.push(path="/etc/config/cert.pem", source=cert_file)
-            self.container.push(path="/etc/config/key.pem", source=key_file)
-            self.container.add_layer("gocert", self._pebble_layer, combine=True)
-            self.container.replan()
+            logger.info("[GoCert] Config file created.")
 
-    def _on_collect_status(self, event: ops.CollectStatusEvent):
-        event.add_status(ops.ActiveStatus())
+    def _configure_access_certificates(self):
+        """Update the config files for gocert and replan if required."""
+        logger.info("[GoCert] Configuring certificates.")
+        certificates_changed = False
+        if not self._self_signed_certificates_generated():
+            certificates_changed = True
+            self._generate_self_signed_certificates()
+        logger.info("[GoCert] Certificates configured.")
```

Comment:
Why the [GoCert] prefix everywhere? Juju will tell us the app/unit that triggered the log

Reply:
The square brackets help me identify the important logs faster because they are easier to see when skimming the sea of logs. It was something we discussed in the pulse retro mentioned by @DanielArndt . I left it because i wanted to get everyone's opinion on it. If we like it i will add it to a custom logger object like we do in Vault, but if we don't I can remove it.

Reply:
I'm ok with changing the way we log but we should write a spec on it as it's quite important and affects every charm. I wouldn't include this change as part of something else though (like here).